tests/test_recommendations.py

```diff
@@ -0,0 +1,204 @@
+        mock_read_ndk.return_value.version = [MAX_NDK_VERSION + 1, 0, 5232133]
+        with self.assertLogs(level="INFO") as cm:
+            check_ndk_version(self.ndk_dir)
+        mock_read_ndk.assert_called_once_with(self.ndk_dir)
+        self.assertEqual(
+            cm.output,
+            [
+                "INFO:p4a:[INFO]:    {}".format(
+                    CURRENT_NDK_VERSION_MESSAGE.format(
+                        ndk_version=MAX_NDK_VERSION + 1
+                    )
+                ),
+                "WARNING:p4a:[WARNING]: {}".format(
+                    RECOMMENDED_NDK_VERSION_MESSAGE.format(
+                        recommended_ndk_version=RECOMMENDED_NDK_VERSION
+                    )
+                ),
+                "WARNING:p4a:[WARNING]: {}".format(NEW_NDK_MESSAGE),
+            ],
+        )
+
+    @mock.patch("pythonforandroid.recommendations.read_ndk_version")
+    def test_check_ndk_version_lower_than_recommended(self, mock_read_ndk):
+        mock_read_ndk.return_value.version = [MIN_NDK_VERSION - 1, 0, 5232133]
+        with self.assertRaises(BuildInterruptingException) as e:
+            check_ndk_version(self.ndk_dir)
+        self.assertEqual(
+            e.exception.args[0],
+            NDK_LOWER_THAN_SUPPORTED_MESSAGE.format(
+                min_supported=MIN_NDK_VERSION, ndk_url=NDK_DOWNLOAD_URL
+            ),
+        )
+        mock_read_ndk.assert_called_once_with(self.ndk_dir)
+
+    @unittest.skipIf(running_in_py2, "`assertLogs` requires Python 3.4+")
+    def test_check_ndk_version_error(self):
+        """
+        Test that a fake ndk dir give us two messages:
+            - first should be an `INFO` log
+            - second should be an `WARNING` log
+        """
+        with self.assertLogs(level="INFO") as cm:
+            check_ndk_version(self.ndk_dir)
+        self.assertEqual(
+            cm.output,
+            [
+                "INFO:p4a:[INFO]:    {}".format(UNKNOWN_NDK_MESSAGE),
+                "WARNING:p4a:[WARNING]: {}".format(
+                    READ_ERROR_NDK_MESSAGE.format(ndk_dir=self.ndk_dir)
+                ),
+                "WARNING:p4a:[WARNING]: {}".format(
+                    ENSURE_RIGHT_NDK_MESSAGE.format(
+                        min_supported=MIN_NDK_VERSION,
+                        rec_version=RECOMMENDED_NDK_VERSION,
+                        ndk_url=NDK_DOWNLOAD_URL,
+                    )
+                ),
+            ],
+        )
+
+    @mock.patch("pythonforandroid.recommendations.open")
+    def test_read_ndk_version(self, mock_open_src_prop):
+        mock_open_src_prop.side_effect = [
+            mock.mock_open(
+                read_data="Pkg.Revision = 17.2.4988734"
+            ).return_value
+        ]
+        version = read_ndk_version(self.ndk_dir)
+        mock_open_src_prop.assert_called_once_with(
+            join(self.ndk_dir, "source.properties")
+        )
+        assert version == "17.2.4988734"
```

Comment:
minor: inconsistent assert. I personally prefer this version, but it's not what you seem to be using in the rest of the file.
Same thing in `test_read_ndk_version_error` below.
But anyway we're far from being consistent in this with p4a currently. Don't know if we want to address or not